Added solutions for exercises

ParnesOmer opened the pull request.

Commit:
Added solutions for exercises

ParnesOmer committed:
Update and rename 5.1.no_vinnigrete.py to 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.4.vessels_communicating.py

ParnesOmer committed:
Update and rename 5.4.vessels_communicating.py to 5.4.communicating_vessels.py

ParnesOmer committed:
Update 5.4.communicating_vessels.py

ParnesOmer committed:
Update 5.4.communicating_vessels.py

ParnesOmer committed:
Update 5.4.communicating_vessels.py

ParnesOmer committed:
Update 5.4.communicating_vessels.py

ParnesOmer committed:
Update and rename 7.2.turtle_sent.py to 7.2.sent_turtle.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.2.cup_of_join.py

ParnesOmer committed:
Update 5.3.parsle_tongue.py

ParnesOmer committed:
Update 5.3.parsle_tongue.py

ParnesOmer committed:
Update 5.3.parsle_tongue.py

ParnesOmer committed:
Add files via upload

ParnesOmer committed:
Update 6.2.running_2000.py

ParnesOmer committed:
Update 6.3.long_cat_is_long.py

ParnesOmer committed:
Update 7.2.sent_turtle.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.thats_the_way.py

ParnesOmer committed:
Update 5.1.thats_the_way.py

ParnesOmer committed:
Update 5.2.cup_of_join.py

ParnesOmer committed:
Update 5.2.piece_of_cake.py

ParnesOmer committed:
Update 5.2.piece_of_cake.py

ParnesOmer committed:
Update 5.2.piece_of_cake.py

ParnesOmer committed:
Update 5.3.parsle_tongue.py

ParnesOmer committed:
Update 5.4.communicating_vessels.py

ParnesOmer committed:
Update 6.2.running_2000.py

ParnesOmer committed:
Update 6.2.running_2000.py

ParnesOmer committed:
Update 6.3.long_cat_is_long.py

ParnesOmer committed:
Update 6.3.long_cat_is_long.py

ParnesOmer committed:
Update 6.4.remember_remember.py

ParnesOmer committed:
Update 6.5.group_by.py

ParnesOmer committed:
Update 6.5.group_by.py

ParnesOmer committed:
Update 7.2.sent_turtle.py

excellenteam-scaleup commented on src/5.1.no_vinnigrette.py:
```diff
@@ -0,0 +1,44 @@
+    random_timestamp = random.randint(0, difference)
+    result = start_date + datetime.timedelta(days=random_timestamp)
+    # Check if the date falls on a Monday
+    if result.weekday() == 0:
```
Do not use "Magic numbers", best practice is to assign to a CONSTANT

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/5.1.no_vinnigrette.py:
```diff
@@ -0,0 +1,44 @@
+    result = start_date + datetime.timedelta(days=random_timestamp)
+    # Check if the date falls on a Monday
+    if result.weekday() == 0:
+        print("Ain't gettin' no vinaigrette today :(")
```
Better to return from function and print from main

excellenteam-scaleup commented on src/5.2.cup_of_join.py:
```diff
@@ -0,0 +1,34 @@
+      Returns:
+      list: A single list containing all elements from input lists, separated by `sep`.
+      """
+    if not lists:
```
Redundant

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/5.2.cup_of_join.py:
```diff
@@ -0,0 +1,34 @@
+        return None  # If no lists are provided, return None
+
+    result = []
+    for i, lst in enumerate(lists):
```
Better to predetermine the separator before loop

excellenteam-scaleup commented on src/5.3.parsle_tongue.py:
```diff
@@ -0,0 +1,57 @@
+    min_size = 5
+    chunk_size = 1024
+    current_string = ""
+    filename = 'logo.jpg'
```
Do not use "Magic values", best practice is to assign to a CONSTANT

excellenteam-scaleup replied:
-3

excellenteam-scaleup commented on src/5.3.parsle_tongue.py:
```diff
@@ -0,0 +1,57 @@
+    chunk_size = 1024
+    current_string = ""
+    filename = 'logo.jpg'
+    try:
```
Redundant try/catch expression... please read: 
https://builtin.com/software-engineering-perspectives/what-is-with-statement-python

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/5.3.parsle_tongue.py:
```diff
@@ -0,0 +1,57 @@
+            while True:
+                chunk = file.read(chunk_size)
+                if not chunk:
+                    break
```
Avoid using break

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/5.4.communicating_vessels.py:
```diff
@@ -0,0 +1,67 @@
+"""
+
+
+def generator_interleave(*iterables):
```
Use list comprehension, leveraging zip_largest

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/5.4.communicating_vessels.py:
```diff
@@ -0,0 +1,67 @@
+                iterators.remove(it)
+
+
+def interleave(*iterables):
```
Use list comprehension, leveraging zip_largest

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/6.2.running_2000.py:
```diff
@@ -0,0 +1,32 @@
+    Returns:
+    float: The execution time in seconds.
+    """
+    t_start = time.time()
```
Why did you decide that it should return the result in seconds? Is it accurate enough for fast functions?

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/6.4.remember_remember.py:
```diff
@@ -0,0 +1,39 @@
+                        encrypted_message.append(chr(h))
+                        break
+            return ''.join(encrypted_message)
+    except FileNotFoundError:
```
Raise

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/6.4.remember_remember.py:
```diff
@@ -0,0 +1,39 @@
+        Returns:
+        str: The decrypted message embedded in the image.
+        """
+    try:
```
Better use with?

excellenteam-scaleup commented on src/6.5.group_by.py:
```diff
@@ -0,0 +1,39 @@
+    dict: A dictionary where the keys are the results of the function f and the values are lists
+          of elements from the iterator that correspond to each key.
+    """
+    try:
```
Redundant try/catch

excellenteam-scaleup replied:
-1

excellenteam-scaleup commented on src/7.2.sent_turtle.py:
```diff
@@ -0,0 +1,89 @@
+        :raises KeyError: if the recipient does not exist.
+        """
+        user_box = self.boxes[recipient]
+        self.message_id = self.message_id + 1
```
Better update counter after actual insertion

excellenteam-scaleup commented on src/5.1.no_vinnigrette.py:
```diff
@@ -0,0 +1,44 @@
+    # Get user input
+    date1 = input("Enter the first date (YYYY-MM-DD): ")
+    date2 = input("Enter the second date (YYYY-MM-DD): ")
+    no_vinnigrete(date1, date2)
```
8/9

excellenteam-scaleup commented on src/5.1.thats_the_way.py:
```diff
@@ -0,0 +1,24 @@
+if __name__ == "__main__":
+    # Check the images folder
+    matching_files = thats_the_way(os.getcwd())
+    print(matching_files)
```
9/9

excellenteam-scaleup commented on src/5.2.cup_of_join.py:
```diff
@@ -0,0 +1,34 @@
+
+if __name__ == "__main__":
+    # Test cases
+    print(cup_of_join([], [1], [], sep='x'))
```
8/9

excellenteam-scaleup commented on src/5.2.piece_of_cake.py:
```diff
@@ -0,0 +1,37 @@
+    # Test cases
+    print(piece_of_cake({'chocolate': 18, 'milk': 8}, chocolate=200, milk=100))  # 44
+    print(piece_of_cake({'chocolate': 18, 'milk': 8}, optionals=['milk'], chocolate=300))  # 54
+    print(piece_of_cake({}))  # 0
```
9/9

excellenteam-scaleup commented on src/5.3.parsle_tongue.py:
```diff
@@ -0,0 +1,57 @@
+    # Using a generator to read the file in parts
+    gen = parsle_tongue()
+    for word in gen:
+        print(word)
```
4/9

excellenteam-scaleup commented on src/5.4.communicating_vessels.py:
```diff
@@ -0,0 +1,67 @@
+    result = list(generator_interleave('abc', [1, 2, 3], ('!', '@', '#')))
+    result2 = interleave('abc', [1, 2, 3], ('!', '@', '#'))
+    print(result)
+    print(result2)
```
7/9

excellenteam-scaleup commented on src/7.2.sent_turtle.py:
```diff
@@ -0,0 +1,89 @@
+            message for message in user_box
+            if search_string in message['body'].lower() or search_string in message['title'].lower()
+        ]
+        return matching_messages
```
9/9

excellenteam-scaleup commented on src/6.5.group_by.py:
```diff
@@ -0,0 +1,39 @@
+    #     return {}
+
+if __name__ == "__main__":
+    print(group_by(len, ["hi", "bye", "yo", "try"]))
```
8/9

excellenteam-scaleup commented on src/6.4.remember_remember.py:
```diff
@@ -0,0 +1,39 @@
+
+
+if __name__ == "__main__":
+    print(remember_remember("code.png"))
```
8/9

excellenteam-scaleup commented on src/6.3.long_cat_is_long.py:
```diff
@@ -0,0 +1,23 @@
+
+if __name__ == '__main__':
+    if long_cat_is_long("bla bla tralala 123") == {'bla': 3, 'tralala': 7}:
+        print('Correct!')
```
9/9

excellenteam-scaleup commented on src/6.2.running_2000.py:
```diff
@@ -0,0 +1,32 @@
+
+
+if __name__ == "__main__":
+    print(running_2000(print, "Hello"))  # Example usage: measuring print execution time
```
8/9

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

I defined a constant named MONDAY at the beginning of the file with the value 0.
I used the constant instead of the value 0 in the code.

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

The function returns a tuple containing the randomly selected date and an indication of whether it is Monday (True/False).
No printing is done within the function, only within the main.

ParnesOmer committed:
Update 5.2.cup_of_join.py

Removing the if not lists test and removing the separator at the end of the list

ParnesOmer committed:
Update 5.2.cup_of_join.py

Pre-defined separator

ParnesOmer committed:
Update 5.3.parsle_tongue.py

Defining constants (MIN_SIZE, CHUNK_SIZE, FILENAME)

ParnesOmer committed:
Update 5.3.parsle_tongue.py

Removing try/except from the parsel_tongue function and adding exception handling outside the function

ParnesOmer committed:
Update 5.4.communicating_vessels.py

use zip_longest

ParnesOmer committed:
Update 6.2.running_2000.py

using rime.pref_counter instead of time.time()

ParnesOmer committed:
Update 6.4.remember_remember.py

Removing the try/except block from the function and adding an UnidentifiedImageError exception

ParnesOmer committed:
Update 6.5.group_by.py

Removing a try/except block

ParnesOmer committed:
Update 7.2.sent_turtle.py

replace thr counter

ParnesOmer committed:
Update 6.4.remember_remember.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.1.no_vinnigrette.py

ParnesOmer committed:
Update 5.2.cup_of_join.py

ParnesOmer committed:
Update 5.2.cup_of_join.py

excellenteam-scaleup reviewed (commented):
+5 bug fixed
